packages/core/src/utils/schema.ts

```diff
@@ -0,0 +1,74 @@
+/**
+ * This module encapsulates format of the path to a deep reference.
+ */
+import { Asset, Entry } from 'contentful';
+
+export type UnresolvedFieldset = Array<[null, string, string?]>;
+export type Fieldset = Array<[Entry | Asset, string, string?]>;
+
+export const parseDataSourcePathIntoFieldset = (deepPath: string): UnresolvedFieldset => {
+  const regex = /^\/([^/]+)\/fields\/(.+)$/; // https://regex101.com/r/cFADIR/1
+
+  const matches = deepPath.match(regex);
+
+  if (null === matches) {
+    throw new Error(
+      `Cannot match deepPath [${deepPath}] with regex [${regex.toString()}] to separate dataSource key from fields.`
+    );
+  }
+
+  const [_match, _key, deepSegments] = matches;
+
+  const segments = deepSegments.split('->');
+  const fieldset: UnresolvedFieldset = segments.map((segment) => {
+    const [field, localeQualifier] = segment.split('/');
+    return [null, field, localeQualifier];
```

Comment:
Indeed, now that you mentioned it, the initial `null` does look odd 🤔  
I have think to make it clearer. 

As to `localeQualifier` i wasn't sure myself what to do with that (as we need to parse it, but then we don't use it 🤷 ).  So do you suggest we just ignore it completely? Aren't we going to lose it when it's used in paths and when it's not used?

Reply:
Re localeQualifier: 
 Now I remember:  the reason I was carrying it around, as at the time of writing this parsing logic, I was expecting that this `~locale` will have meaning. As the path can be defined with or without it eg. 

```
# From what I understood such path will always "fill in" the actual locale
# for both `refPrimaryImage` and for the Asset (`file`) 
/key123/fields/refPrimaryImage/~locale->file/~locale
```

Whereas such path

```
# From what I understand such file will always fall back to default locale
/key123/fields/refPrimaryImage->file

# And such path would "fill in" locale only for `refPrimaryImage` , but not for Asset(`file`)
/key123/fields/refPrimaryImage/~locale->file` 
```

But later, after looking at the SDK code, It turned out that `~locale` is pretty much ignored. 🤷  (i grepped for `~locale` and for `split('/')` and at those sites locale is never used)

Reply:
Is there a case where the locale is not used at all? As far as I know when resolving a field value you always have to select the value of a locale. So will we ever have `/key123/fields/refPrimaryImage/~locale->file`? I don't think so.
The locale at the moment is just a placeholder, so I think it can be ignored. Maybe later we'll support specifying a specific locale, e.g. `/key123/fields/refPrimaryImage/en-US` but I'd leave this out of scope for now.

Reply:
I have rewritten this logic, to assume (as per your suggestion) that `/fields/fieldName/~locale` is the format (and that `~locale` is always present in every segment of the field path.